### PR TITLE
Be able to connect to store with arbitrary socket settings

### PR DESCRIPTION
**Pull request by thomasjm**
I wanted to be able to connect to a Nix store over TCP, rather than using a Unix socket. (This is pretty easy to set up using the technique described [here](https://github.com/NixOS/nix/issues/830).)

This PR exposed a new function `runStoreOptsTCP`, which allows you to specify a hostname and port rather than a Unix socket path. It also exposes an even lower-level function `runStoreOptsInner` which allows you to use arbitrary socket settings.

### CODE REVIEW

File: hnix-store-remote/src/System/Nix/Store/Remote/Protocol.hs [hunk 1]
```diff
@@ -168,17 +171,25 @@ runStore = runStoreOpts defaultSockPath "/nix/store"
...
 runStoreOpts
   :: FilePath -> FilePath -> MonadStore a -> IO (Either String a, [Logger])
-runStoreOpts sockPath storeRootDir code = do
-  bracket (open sockPath) (Network.Socket.close . storeSocket) run
+runStoreOpts path = runStoreOptsInner S.AF_UNIX (SockAddrUnix path)
+
+runStoreOptsTCP
+  :: String -> Int -> FilePath -> MonadStore a -> IO (Either String a, [Logger])
+runStoreOptsTCP host port storeRootDir code = do
+  let hints = S.defaultHints
```
Comment (Anton-Latukha) [inline hnix-store-remote/src/System/Nix/Store/Remote/Protocol.hs:179]: No point in creating additional abstraction here: https://wiki.haskell.org/Fairbairn_threshold

File: hnix-store-remote/src/System/Nix/Store/Remote/Protocol.hs [hunk 2]
```diff
@@ -168,17 +171,25 @@ runStore = runStoreOpts defaultSockPath "/nix/store"
...
+  S.getAddrInfo (Just hints) (Just host) (Just $ show port) >>= \case
+    (sockAddr:_) -> runStoreOptsInner (S.addrFamily sockAddr) (S.addrAddress sockAddr) storeRootDir code
+    _ -> return (Left "Couldn't resolve host and port with getAddrInfo.", [])
```
Comment (Anton-Latukha) [inline hnix-store-remote/src/System/Nix/Store/Remote/Protocol.hs:182]: `return` entered deprecation in GHC 9.2 https://github.com/fumieval/Haskell-breaking-changes#planned-remove-return-from-monad.

**Comment by Anton-Latukha** [comment]
The above are just comments they are not needed to be fixed.

---

PR is not rebased onto the project branch, which prevents PR to be merged.

**Comment by thomasjm** [comment]
Rebased now

**Comment by Anton-Latukha** [comment]
My notes of course a bit of [`Wadlet's Law`](https://wiki.haskell.org/Wadler%27s_Law).

Thank you 8)

**Pull request merged**